mycompany-seeq-link-connector-myconnector/src/main/java/com/mycompany/seeq/link/connector/DatasourceSimulator.java

```diff
@@ -30,55 +94,177 @@ public String getName() {
+    public void disconnect() {}
+
+    public Iterable<Element> getDatabases() {
+        int databaseCount = RNG.nextInt(10);
+        return IntStream.range(1, databaseCount + 1)
+                .mapToObj(Element::new)
+                .collect(Collectors.toList());
     }
 
-    public void disconnect() {}
+    public Iterable<Alarm> getAlarmsForDatabase(String elementId) {
+        int alarmCount = RNG.nextInt(10);
+        return IntStream.range(1, alarmCount + 1)
+                .mapToObj(alarmId -> new Alarm(elementId, alarmId))
+                .collect(Collectors.toList());
+    }
+
+    public Iterable<Tag> getTagsForDatabase(String elementId) {
+        int tagCount = RNG.nextInt(10);
+        return IntStream.range(1, tagCount + 1)
+                .mapToObj(tagId -> new Tag(elementId, tagId, tagId % 2 == 0))
+                .collect(Collectors.toList());
+    }
+
+    public Iterable<Constant> getConstantsForDatabase(String elementId) {
+        int constantCount = RNG.nextInt(10);
+        return IntStream.range(1, constantCount + 1)
+                .mapToObj(constId -> new Constant(elementId, constId, "°C", constId * 10))
+                .collect(Collectors.toList());
+    }
 
-    public int getTagCount() {
-        return this.tagCount;
+    public enum Waveform {
+        SINE
     }
 
-    public Iterator<Tag> getTags() {
-        return new Iterator<Tag>() {
-            private int nextTagNumber = 0;
+    public Iterable<Tag.Value> getTagValues(String dataId, TimeInstant startTimestamp, TimeInstant endTimestamp,
+            int limit) {
+        long samplePeriodInNanos = this.signalPeriod.toNanos();
+        long leftBoundTimestamp = startTimestamp.getTimestamp() / samplePeriodInNanos;
+        long rightBoundTimestamp = (endTimestamp.getTimestamp() + samplePeriodInNanos - 1) / samplePeriodInNanos;
+
+        return () -> new Iterator<>() {
+            private long sampleIndex = leftBoundTimestamp;
+            private long count = 0;
 
             @Override
             public boolean hasNext() {
-                return (this.nextTagNumber < DatasourceSimulator.this.tagCount);
+                return sampleIndex <= rightBoundTimestamp && count < limit;
             }
 
             @Override
-            public Tag next() {
-                this.nextTagNumber++;
-                return new Tag(
-                        this.nextTagNumber,
-                        String.format("Simulated Tag #%d", this.nextTagNumber),
-                        this.nextTagNumber % 2 == 0);
+            public Tag.Value next() {
+                if (!hasNext()) {
+                    throw new RuntimeException();
+                }
+
+                TimeInstant key = new TimeInstant(sampleIndex * samplePeriodInNanos);
+                double value = getWaveformValue(Waveform.SINE, key.getTimestamp());
+                sampleIndex++;
+                count++;
+                return new Tag.Value(key, value);
```

Comment:
Maybe you could avoid implementing an iterator by using a stream, something like (untested):

```
        LongStream.range(0, Long.MAX_VALUE)
                .mapToObj(i -> {
                        TimeInstant key = new TimeInstant(i * samplePeriodInNanos);
                        double value = getWaveformValue(Waveform.SINE, key.getTimestamp());
                        return new Tag.Value(key, value);
                }).limit(limit)
                .iterator()
```

Reply:
I tend to like the stream convention more, as the link sdk is stream oriented